File: python/fiesta/src/part2grid_wei.f90

subroutine weight_cic(xp, xg, dl, w)

  ! Cloud-in-Cell weigths in 1D.
  !
  ! Parameters
  ! ----------
  ! xp : float
  !   Coordinate of the particle.
  ! xg : float
  !   Coordinate of the grid point.
  ! dl : float
  !   Size of the cell.
  !
  ! Returns
  ! -------
  ! w : float
  !   Weight

  implicit none
  integer, parameter :: dp = kind(1.d0)

  ! define variales.

  real(kind=dp), intent(in) :: xp, xg, dl
  real(kind=dp), intent(out) :: w

  real(kind=dp) :: dx

  ! absolute difference between particle and grid point.

  dx = abs(xp - xg)

  ! calculating weight

  if (dx .LE. dl) then
    w = (1. - (dx/dl))/dl
  else
    w = 0.
  end if

end subroutine weight_cic

subroutine weight_tsc(xp, xg, dl, w)

  ! Triangular-shaped-cloud weigths in 1D.
  !
  ! Parameters
  ! ----------
  ! xp : float
  !   Coordinate of the particle.
  ! xg : float
  !   Coordinate of the grid point.
  ! dl : float
  !   Size of the cell.
  !
  ! Returns
  ! -------
  ! w : float
  !   Weight

  implicit none
  integer, parameter :: dp = kind(1.d0)

  ! define variables

  real(kind=dp), intent(in) :: xp, xg, dl
  real(kind=dp), intent(out) :: w

  real(kind=dp) :: dx

  ! absolute difference between particle and grid point.

  dx = abs(xp - xg)

  ! calculating weight

  if (dx .LE. dl/2.) then
    w = 3./4. - (dx/dl)**2.
  else if (dx .LE. 3.*dl/2.) then
    w = (1./2.)*((3./2. - (dx/dl))**2.)
  else
    w = 0.
  end if

  w = w / dl

end subroutine weight_tsc

File: fiesta/src/part2grid_wei.py
import numpy as np
from numba import njit


@njit
def weight_cic(xp, xg, dl):
    """
    Cloud-in-Cell (CIC) weights in 1D.
    
    Parameters:
    xp : float
        Coordinate of the particle.
    xg : float
        Coordinate of the grid point.
    dl : float
        Size of the grid cell.

    Returns:
    w : float
        Weight computed based on the distance between the particle and the grid point.
    """
    # Compute the absolute distance between the particle and the grid point
    dx = abs(xp - xg)
    
    # Calculate the weight using the Cloud-in-Cell method
    if dx <= dl:
        w = (1. - (dx / dl)) / dl
    else:
        w = 0.0
    
    return w


@njit
def weight_tsc(xp, xg, dl):
    """
    Triangular-shaped-cloud (TSC) weights in 1D.
    
    Parameters:
    xp : float
        Coordinate of the particle.
    xg : float
        Coordinate of the grid point.
    dl : float
        Size of the grid cell.
    
    Returns:
    w : float
        Weight computed based on the triangular-shaped cloud function.
    """
    # Compute the absolute distance between the particle and the grid point
    dx = abs(xp - xg)
    
    # Calculate the weight using the Triangular-Shaped-Cloud method
    if dx <= dl / 2.:
        w = 3. / 4. - (dx / dl) ** 2.
    elif dx <= 3. * dl / 2.:
        w = (1./2.) * ((3. / 2. - (dx / dl)) ** 2.)
    else:
        w = 0.0
    
    # Normalize the weight by the cell size
    w = w / dl
    
    return w


@njit
def weight_pcs(xp, xg, dl):
    """
    Piecewise-Cubic-Spline (PCS) weights in 1D.
    
    Parameters:
    xp : float
        Coordinate of the particle.
    xg : float
        Coordinate of the grid point.
    dl : float
        Size of the grid cell.
    
    Returns:
    w : float
        Weight computed based on the Piecewise-Cubic-Spline function.
    """
    # Compute the absolute distance between the particle and the grid point
    dx = abs(xp - xg)
    
    # Calculate the weight using the Piecewise-Cubic-Spline method
    if dx <= dl:
        w = (1. / 6.) * (4. - 6. * (dx / dl) ** 2. + 3. * (dx / dl) ** 3.)
    elif dx <= 2. * dl:
        w = (1. / 6.) * ((2. - (dx / dl)) ** 3.)
    else:
        w = 0.0
    
    # Normalize the weight by the cell size
    w = w / dl
    
    return w
